src/Storages/StorageS3.cpp

```diff
@@ -573,7 +573,10 @@ StorageS3Source::StorageS3Source(
 {
     reader = createReader();
     if (reader)
+    {
+        total_objects_size = tryGetFileSizeFromReadBuffer(*reader.getReadBuffer()).value_or(0);
```

Comment:
so it will mean that we do head requests in constructor of storageS3Source? may be delay it until actually needed? It seems to be more optimal

Reply:
Interesting, why not delay creating a reader then.. but not related to this pr anyway